python/vsi/vendored/filelock.py

```diff
@@ -0,0 +1,459 @@
+
+    BaseFileLock._locks.add(self)
+
+    return None
+
+  @property
+  def lock_file(self):
+    """
+    The path to the lock file.
+    """
+    return self._lock_file
+
+  @property
+  def timeout(self):
+    """
+    You can set a default timeout for the filelock. It will be used as
+    fallback value in the acquire method, if no timeout value (*None*) is
+    given.
+
+    If you want to disable the timeout, set it to a negative value.
+
+    A timeout of 0 means, that there is exactly one attempt to acquire the
+    file lock.
+
+    .. versionadded:: 2.0.0
+    """
+    return self._timeout
+
+  @timeout.setter
+  def timeout(self, value):
+    """
+    """
+    self._timeout = float(value)
+    return None
+
+  # Platform dependent locking
+  # --------------------------------------------
+
+  def _acquire(self):
+    """
+    Platform dependent. If the file lock could be
+    acquired, self._lock_file_fd holds the file descriptor
+    of the lock file.
+    """
+    raise NotImplementedError()
+
+  def _release(self):
+    """
+    Releases the lock and sets self._lock_file_fd to None.
+    """
+    raise NotImplementedError()
+
+  # Platform independent methods
+  # --------------------------------------------
+
+  @property
+  def is_locked(self):
+    """
+    True, if the object holds the file lock.
+
+    .. versionchanged:: 2.0.0
+
+      This was previously a method and is now a property.
+    """
+    return self._lock_file_fd is not None
+
+  def acquire(self, timeout=None, poll_intervall=0.05):
+    """
+    Acquires the file lock or fails with a :exc:`Timeout` error.
+
+    .. code-block:: python
+
+      # You can use this method in the context manager (recommended)
+      with lock.acquire():
+        pass
+
+      # Or use an equivalent try-finally construct:
+      lock.acquire()
+      try:
+        pass
+      finally:
+        lock.release()
+
+    :arg float timeout:
+      The maximum time waited for the file lock.
+      If ``timeout < 0``, there is no timeout and this method will
+      block until the lock could be acquired.
+      If ``timeout`` is None, the default :attr:`~timeout` is used.
+
+    :arg float poll_intervall:
+      We check once in *poll_intervall* seconds if we can acquire the
+      file lock.
+
+    :raises Timeout:
+      if the lock could not be acquired in *timeout* seconds.
+
+    .. versionchanged:: 2.0.0
+
+      This method returns now a *proxy* object instead of *self*,
+      so that it can be used in a with statement without side effects.
+    """
+    # Use the default timeout, if no timeout is provided.
+    if timeout is None:
+      timeout = self.timeout
+
+    # Increment the number right at the beginning.
+    # We can still undo it, if something fails.
+    with self._thread_lock:
+      self._lock_counter += 1
+
+    lock_id = id(self)
+    lock_filename = self._lock_file
+    start_time = time.time()
+    try:
+      while True:
+        with self._thread_lock:
```

Comment:
I'm surprised that it is necessary to acquire a (thread) lock here. I would expect the filelock to be thread-safe i guess